sopel/tools/calculation.py

```diff
@@ -57,8 +57,11 @@ def _eval_node(self, node, timeout):
         A subclass could overwrite this to handle more nodes, calling it only
         for nodes it does not implement itself.
         """
-        if isinstance(node, ast.Num):
-            return node.n
+        if (
+            isinstance(node, ast.Constant) and
+            isinstance(node.value, (int, float))
```

Comment:
I think we need to handle `complex` here as well?

```suggestion
            isinstance(node.value, (int, float, complex))
```

Reply:
Ha, I had that in an earlier revision. Then I realized that `complex` being handled might be an accident, and took it out.

What do you think of the performance implications if we do leave `complex` as one of the supported types? Given the whole point of this expression evaluator class is preventing `.calc` input from pegging a whole CPU thread for more than a couple seconds.

Reply:
> What do you think of the performance implications if we do leave `complex` as one of the supported types?

"Measure it" is usually what I think about perf :stuck_out_tongue_winking_eye:

But the `isinstance` check should be pretty snappy, I don't think including one more type would be a problem.

Reply:
My worry was about the performance of operations that _use_ `complex` values, but it doesn't seem to be an issue. `timeit` tests showed about 4 ns per loop for `3 * 3` and 19 ns per loop for `3 * c` (using `--setup 'c = complex(3, 5)` to exclude the object initialization from timing measurements). Doesn't seem like a big enough difference to specifically disallow the use of `complex` values.

Reply:
Ooh, however, the real reason for not supporting `complex` is that `guarded_mul()` and `guarded_pow()` aren't written to support it. They explicitly accept only `int` or `float` values.